File: android/app/src/main/java/com/kordar/ghostchat/core/files/FileTransferService.kt
package com.kordar.ghostchat.core.files

import com.kordar.ghostchat.models.ControlMessage
import java.security.MessageDigest
import java.util.UUID

/**
 * Pure state machine for chunked file transfer over the encrypted DataChannel.
 * Produces [ControlMessage] payloads for sending; reassembles them on the receive
 * side, verifying SHA-256 integrity.
 *
 * Does NOT encrypt or touch the network — `ConnectionManager` wraps each produced
 * [ControlMessage] with `GhostChatCrypto.encrypt` and sends it through the WebRTC
 * data channel, applying backpressure.
 *
 * Mirror of iOS `Core/Files/FileTransferService.swift` — both platforms must
 * produce byte-identical chunks + hash for identical inputs.
 */
class FileTransferService {

    companion object {
        /** Raw bytes per chunk. Must match iOS. */
        const val CHUNK_SIZE: Int = 2048

        /** Android uses `java.util.Base64` (API 26+, our minSdk is 28) to avoid
         *  the URL-safe variant that `android.util.Base64.encodeToString(..., Base64.DEFAULT)`
         *  produces and to stay byte-identical with iOS `Data.base64EncodedString()`. */
        internal val base64Encoder = java.util.Base64.getEncoder()
        internal val base64Decoder = java.util.Base64.getDecoder()
    }

    data class Outbound(
        val fileId: String,
        val name: String,
        val size: Int,
        val mimeType: String,
        val sha256Hex: String,
        val totalChunks: Int,
        val startMessage: ControlMessage,
        val chunkMessages: List<ControlMessage>,
        val completeMessage: ControlMessage
    )

    data class IncomingFile(
        val fileId: String,
        val name: String,
        val mimeType: String,
        val data: ByteArray
    ) {
        override fun equals(other: Any?): Boolean {
            if (this === other) return true
            if (other !is IncomingFile) return false
            return fileId == other.fileId && name == other.name &&
                    mimeType == other.mimeType && data.contentEquals(other.data)
        }
        override fun hashCode(): Int {
            var r = fileId.hashCode()
            r = 31 * r + name.hashCode()
            r = 31 * r + mimeType.hashCode()
            r = 31 * r + data.contentHashCode()
            return r
        }
    }

    sealed class Event {
        data class Started(val fileId: String) : Event()
        data class Progressed(val fileId: String, val received: Int, val total: Int) : Event()
        data class Completed(val file: IncomingFile) : Event()
        data class IntegrityFailure(val fileId: String) : Event()
        data class Missing(val fileId: String, val indices: List<Int>) : Event()
        data class Unknown(val fileId: String) : Event()
    }

    private data class Inbound(
        val name: String,
        val size: Int,
        val mimeType: String,
        val totalChunks: Int,
        val chunks: MutableMap<Int, ByteArray> = mutableMapOf(),
        val startedAtMs: Long,
        var lastChunkAtMs: Long
    )

    private data class OutboundState(
        val totalChunks: Int,
        val chunkMessages: List<ControlMessage>
    )

    private val inbound = mutableMapOf<String, Inbound>()
    private val outbound = mutableMapOf<String, OutboundState>()

    // MARK: - Outbound

    fun prepareOutbound(data: ByteArray, name: String, mimeType: String): Outbound {
        require(name.isNotEmpty()) { "file name must not be empty" }

        val fileId = UUID.randomUUID().toString()
        val sha256 = sha256Hex(data)
        val chunks = chunked(data, CHUNK_SIZE)
        val start = ControlMessage.FileStart(
            fileId = fileId,
            name = name,
            size = data.size,
            mimeType = mimeType,
            totalChunks = chunks.size
        )
        val chunkMessages: List<ControlMessage> = chunks.mapIndexed { index, raw ->
            ControlMessage.FileChunk(
                fileId = fileId,
                index = index,
                data = base64Encoder.encodeToString(raw)
            )
        }
        val complete = ControlMessage.FileComplete(fileId = fileId, sha256 = sha256)
        outbound[fileId] = OutboundState(totalChunks = chunks.size, chunkMessages = chunkMessages)
        return Outbound(
            fileId = fileId,
            name = name,
            size = data.size,
            mimeType = mimeType,
            sha256Hex = sha256,
            totalChunks = chunks.size,
            startMessage = start,
            chunkMessages = chunkMessages,
            completeMessage = complete
        )
    }

    // MARK: - Inbound

    fun handleStart(
        fileId: String,
        name: String,
        size: Int,
        mimeType: String,
        totalChunks: Int
    ): Event {
        val now = System.currentTimeMillis()
        inbound[fileId] = Inbound(
            name = name, size = size, mimeType = mimeType,
            totalChunks = totalChunks,
            startedAtMs = now, lastChunkAtMs = now
        )
        return Event.Started(fileId)
    }

    fun handleChunk(fileId: String, index: Int, base64Data: String): Event {
        val entry = inbound[fileId] ?: return Event.Unknown(fileId)
        if (index < 0 || index >= entry.totalChunks) return Event.Unknown(fileId)
        val bytes = try {
            base64Decoder.decode(base64Data)
        } catch (_: IllegalArgumentException) {
            return Event.Unknown(fileId)
        }
        entry.chunks[index] = bytes
        entry.lastChunkAtMs = System.currentTimeMillis()
        return Event.Progressed(fileId, received = entry.chunks.size, total = entry.totalChunks)
    }

    fun handleComplete(fileId: String, expectedSha256Hex: String): Event {
        val entry = inbound[fileId] ?: return Event.Unknown(fileId)

        val missing = (0 until entry.totalChunks).filter { !entry.chunks.containsKey(it) }
        if (missing.isNotEmpty()) {
            return Event.Missing(fileId, missing)
        }

        val assembled = ByteArray(entry.size)
        var offset = 0
        for (i in 0 until entry.totalChunks) {
            val c = entry.chunks[i] ?: continue
            System.arraycopy(c, 0, assembled, offset, c.size)
            offset += c.size
        }

        val actual = sha256Hex(assembled)
        if (!actual.equals(expectedSha256Hex, ignoreCase = true)) {
            inbound.remove(fileId)
            return Event.IntegrityFailure(fileId)
        }

        inbound.remove(fileId)
        return Event.Completed(
            IncomingFile(
                fileId = fileId,
                name = entry.name,
                mimeType = entry.mimeType,
                data = assembled
            )
        )
    }

    fun missingChunks(fileId: String): List<Int>? {
        val entry = inbound[fileId] ?: return null
        return (0 until entry.totalChunks).filter { !entry.chunks.containsKey(it) }
    }

    fun cancelInbound(fileId: String) {
        inbound.remove(fileId)
    }

    fun hasInbound(fileId: String): Boolean = inbound.containsKey(fileId)

    /** Re-emit previously-built fileChunk messages for a retransmit request. */
    fun retransmitMessages(fileId: String, indices: List<Int>): List<ControlMessage> {
        val state = outbound[fileId] ?: return emptyList()
        return indices.mapNotNull { i ->
            if (i in 0 until state.chunkMessages.size) state.chunkMessages[i] else null
        }
    }

    fun forgetOutbound(fileId: String) {
        outbound.remove(fileId)
    }

    fun hasOutbound(fileId: String): Boolean = outbound.containsKey(fileId)

    // MARK: - Helpers

    private fun chunked(data: ByteArray, size: Int): List<ByteArray> {
        if (data.isEmpty()) return emptyList()
        val out = ArrayList<ByteArray>((data.size + size - 1) / size)
        var offset = 0
        while (offset < data.size) {
            val end = (offset + size).coerceAtMost(data.size)
            out += data.copyOfRange(offset, end)
            offset = end
        }
        return out
    }

    private fun sha256Hex(data: ByteArray): String {
        val digest = MessageDigest.getInstance("SHA-256").digest(data)
        val sb = StringBuilder(digest.size * 2)
        for (b in digest) sb.append("%02x".format(b.toInt() and 0xFF))
        return sb.toString()
    }
}

File: ios/GhostChat/Core/Files/FileTransferService.swift
import CryptoKit
import Foundation

/// Errors raised by `FileTransferService.prepareOutbound(...)`.
enum FileTransferError: Error, Equatable {
    case emptyName
}

/// Pure state machine for chunked file transfer over the encrypted DataChannel.
/// Produces `ControlMessage` payloads for sending; reassembles them on the
/// receive side, verifying SHA-256 integrity.
///
/// Does NOT encrypt or touch the network — `ConnectionManager` wraps each
/// produced `ControlMessage` with `GhostChatCrypto.encrypt` and sends it
/// through the WebRTC data channel, applying backpressure.
///
/// Mirror of Android `core/files/FileTransferService.kt` — both platforms must
/// produce byte-identical chunks + hash for identical inputs.
final class FileTransferService {

    /// Raw bytes per chunk. Base64 overhead (~33%) + Double Ratchet header keeps
    /// each encrypted frame well under 4 KiB on the wire.
    static let chunkSize = 2048

    struct Outbound: Equatable {
        let fileId: String
        let name: String
        let size: Int
        let mimeType: String
        let sha256Hex: String
        let totalChunks: Int
        let startMessage: ControlMessage
        let chunkMessages: [ControlMessage]
        let completeMessage: ControlMessage
    }

    struct IncomingFile: Equatable {
        let fileId: String
        let name: String
        let mimeType: String
        let data: Data
    }

    enum Event: Equatable {
        case started(fileId: String)
        case progressed(fileId: String, received: Int, total: Int)
        case completed(IncomingFile)
        case integrityFailure(fileId: String)
        case missing(fileId: String, indices: [Int])
        case unknown(fileId: String)
    }

    private struct Inbound {
        let name: String
        let size: Int
        let mimeType: String
        let totalChunks: Int
        var chunks: [Int: Data] = [:]
        var startedAt: Date
        var lastChunkAt: Date
    }

    private struct OutboundState {
        let totalChunks: Int
        let chunkMessages: [ControlMessage]
    }

    private var inbound: [String: Inbound] = [:]
    private var outbound: [String: OutboundState] = [:]

    // MARK: - Outbound

    func prepareOutbound(data: Data, name: String, mimeType: String) throws -> Outbound {
        guard !name.isEmpty else { throw FileTransferError.emptyName }

        let fileId = UUID().uuidString
        let sha256Hex = Self.sha256Hex(data)
        let chunks = Self.chunked(data: data, size: Self.chunkSize)
        let start: ControlMessage = .fileStart(
            fileId: fileId,
            name: name,
            size: data.count,
            mimeType: mimeType,
            totalChunks: chunks.count
        )
        let chunkMessages: [ControlMessage] = chunks.enumerated().map { index, raw in
            .fileChunk(fileId: fileId, index: index, data: raw.base64EncodedString())
        }
        let complete: ControlMessage = .fileComplete(fileId: fileId, sha256: sha256Hex)
        outbound[fileId] = OutboundState(totalChunks: chunks.count, chunkMessages: chunkMessages)

        return Outbound(
            fileId: fileId,
            name: name,
            size: data.count,
            mimeType: mimeType,
            sha256Hex: sha256Hex,
            totalChunks: chunks.count,
            startMessage: start,
            chunkMessages: chunkMessages,
            completeMessage: complete
        )
    }

    // MARK: - Inbound

    @discardableResult
    func handleStart(fileId: String, name: String, size: Int, mimeType: String, totalChunks: Int) -> Event {
        let now = Date()
        inbound[fileId] = Inbound(
            name: name, size: size, mimeType: mimeType,
            totalChunks: totalChunks,
            startedAt: now, lastChunkAt: now
        )
        return .started(fileId: fileId)
    }

    @discardableResult
    func handleChunk(fileId: String, index: Int, base64Data: String) -> Event {
        guard var entry = inbound[fileId] else { return .unknown(fileId: fileId) }
        guard index >= 0, index < entry.totalChunks else { return .unknown(fileId: fileId) }
        guard let bytes = Data(base64Encoded: base64Data) else { return .unknown(fileId: fileId) }

        entry.chunks[index] = bytes
        entry.lastChunkAt = Date()
        inbound[fileId] = entry
        return .progressed(fileId: fileId, received: entry.chunks.count, total: entry.totalChunks)
    }

    @discardableResult
    func handleComplete(fileId: String, expectedSha256Hex: String) -> Event {
        guard let entry = inbound[fileId] else { return .unknown(fileId: fileId) }

        let missing = (0..<entry.totalChunks).filter { entry.chunks[$0] == nil }
        if !missing.isEmpty {
            return .missing(fileId: fileId, indices: missing)
        }

        var data = Data()
        data.reserveCapacity(entry.size)
        for i in 0..<entry.totalChunks {
            if let chunk = entry.chunks[i] { data.append(chunk) }
        }

        let actual = Self.sha256Hex(data)
        if actual.lowercased() != expectedSha256Hex.lowercased() {
            inbound.removeValue(forKey: fileId)
            return .integrityFailure(fileId: fileId)
        }

        inbound.removeValue(forKey: fileId)
        return .completed(IncomingFile(
            fileId: fileId,
            name: entry.name,
            mimeType: entry.mimeType,
            data: data
        ))
    }

    func missingChunks(fileId: String) -> [Int]? {
        guard let entry = inbound[fileId] else { return nil }
        return (0..<entry.totalChunks).filter { entry.chunks[$0] == nil }
    }

    func cancelInbound(fileId: String) {
        inbound.removeValue(forKey: fileId)
    }

    func hasInbound(fileId: String) -> Bool {
        inbound[fileId] != nil
    }

    /// Re-emit previously-built `fileChunk` messages for a retransmit request.
    /// Returns an empty array if the outbound file has already been forgotten.
    func retransmitMessages(fileId: String, indices: [Int]) -> [ControlMessage] {
        guard let state = outbound[fileId] else { return [] }
        return indices.compactMap { i in
            (i >= 0 && i < state.chunkMessages.count) ? state.chunkMessages[i] : nil
        }
    }

    /// Drop the retained outbound chunks; called by `ConnectionManager`
    /// after the transfer has completed on both sides.
    func forgetOutbound(fileId: String) {
        outbound.removeValue(forKey: fileId)
    }

    func hasOutbound(fileId: String) -> Bool {
        outbound[fileId] != nil
    }

    // MARK: - Helpers

    private static func chunked(data: Data, size: Int) -> [Data] {
        guard !data.isEmpty else { return [] }
        var result: [Data] = []
        result.reserveCapacity((data.count + size - 1) / size)
        var offset = 0
        while offset < data.count {
            let end = min(offset + size, data.count)
            result.append(data.subdata(in: offset..<end))
            offset = end
        }
        return result
    }

    private static func sha256Hex(_ data: Data) -> String {
        let digest = SHA256.hash(data: data)
        return digest.map { String(format: "%02x", $0) }.joined()
    }
}
